src/anndata/_core/merge.py

```diff
@@ -647,21 +750,56 @@ def _apply_to_cupy_array(self, el, *, axis, fill_value=None):
 
         return out
 
-    def _apply_to_array(self, el, *, axis, fill_value=None):
+    def _apply_to_array_api(self, el, *, axis, fill_value=None):
         if fill_value is None:
             fill_value = default_fill_value([el])
-        if el.shape[axis] == 0:
-            # Presumably faster since it won't allocate the full array
-            shape = list(el.shape)
-            shape[axis] = len(self.new_idx)
-            return np.broadcast_to(fill_value, tuple(shape))
 
         indexer = self.idx
 
-        # Indexes real fast, and does outer indexing
-        return pd.api.extensions.take(
-            el, indexer, axis=axis, allow_fill=True, fill_value=fill_value
-        )
+        if _is_pandas(el):
```

Comment:
```suggestion
        if isinstance(el, np.ndarray):
```
I would have thought that `el` is a numpy array given that the old function name was `_apply_to_array`, no?